File: spec/javascript/server/handshake_spec.js
var jstest = require("jstest").Test

var Engine = require("../../../javascript/engines/proxy"),
    Server = require("../../../javascript/protocol/server")

jstest.describe("Server handshake", function() { with(this) {
  before(function() { with(this) {
    this.engine = {}
    stub(Engine, "get").returns(engine)
    this.server = new Server()

    this.connectionTypes = ["long-polling", "cross-origin-long-polling",
                            "callback-polling","websocket",
                            "eventsource","in-process"]
  }})

  describe("#handshake", function() { with(this) {
    before(function() { with(this) {
      this.message = {channel: "/meta/handshake",
                      version: "1.0",
                      supportedConnectionTypes: ["long-polling"]}
    }})

    describe("with valid parameters", function() { with(this) {
      it("creates a client", function() { with(this) {
        expect(engine, "createClient")
        server.handshake(message, false, function() {})
      }})

      it("returns a successful response", function() { with(this) {
        stub(engine, "createClient").yields(["clientid"])
        server.handshake(message, false, function(response) {
          assertEqual({
              channel:    "/meta/handshake",
              successful: true,
              version:    "1.0",
              supportedConnectionTypes: connectionTypes,
              clientId:   "clientid"
            }, response)
        })
      }})

      describe("with a message id", function() { with(this) {
        before(function() { this.message.id = "foo" })

        it("returns the same id", function() { with(this) {
          stub(engine, "createClient").yields(["clientid"])
          server.handshake(message, false, function(response) {
            assertEqual({
                channel:    "/meta/handshake",
                successful: true,
                version:    "1.0",
                supportedConnectionTypes: connectionTypes,
                clientId:   "clientid",
                id:         "foo"
              }, response)
          })
        }})
      }})
    }})

    describe("missing version", function() { with(this) {
      before(function() { delete this.message.version })

      it("does not create a client", function() { with(this) {
        expect(engine, "createClient").exactly(0)
        server.handshake(message, false, function() {})
      }})

      it("returns an unsuccessful response", function() { with(this) {
        server.handshake(message, false, function(response) {
          assertEqual({
              channel:    "/meta/handshake",
              successful: false,
              error:      "402:version:Missing required parameter",
              version:    "1.0",
              supportedConnectionTypes: connectionTypes
            }, response)
        })
      }})
    }})

    describe("missing supportedConnectionTypes", function() { with(this) {
      before(function() { delete this.message.supportedConnectionTypes })

      it("does not create a client", function() { with(this) {
        expect(engine, "createClient").exactly(0)
        server.handshake(message, false, function() {})
      }})

      it("returns an unsuccessful response", function() { with(this) {
        server.handshake(message, false, function(response) {
          assertEqual({
              channel:    "/meta/handshake",
              successful: false,
              error:      "402:supportedConnectionTypes:Missing required parameter",
              version:    "1.0",
              supportedConnectionTypes: connectionTypes
            }, response)
        })
      }})
    }})

    describe("with no matching supportedConnectionTypes", function() { with(this) {
      before(function() { with(this) {
        message.supportedConnectionTypes = ["iframe", "flash"]
      }})

      it("does not create a client", function() { with(this) {
        expect(engine, "createClient").exactly(0)
        server.handshake(message, false, function() {})
      }})

      it("returns an unsuccessful response", function() { with(this) {
        server.handshake(message, false, function(response) {
          assertEqual({
              channel:    "/meta/handshake",
              successful: false,
              error:      "301:iframe,flash:Connection types not supported",
              version:    "1.0",
              supportedConnectionTypes: connectionTypes
            }, response)
        })
      }})
    }})

    describe("with an error", function() { with(this) {
      before(function() { with(this) {
        message.error = "invalid"
      }})

      it("does not create a client", function() { with(this) {
        expect(engine, "createClient").exactly(0)
        server.handshake(message, false, function() {})
      }})

      it("returns an unsuccessful response", function() { with(this) {
        server.handshake(message, false, function(response) {
          assertEqual({
              channel:    "/meta/handshake",
              successful: false,
              error:      "invalid",
              version:    "1.0",
              supportedConnectionTypes: connectionTypes
            }, response)
        })
      }})
    }})
  }})
}})

File: spec/ruby/server/handshake_spec.rb
require "spec_helper"

describe "server handshake" do
  let(:engine) { double "engine" }
  let(:server) { Faye::Server.new }

  let :connection_types do
    ["long-polling","cross-origin-long-polling","callback-polling","websocket","eventsource","in-process"]
  end

  before do
    Faye::Engine.stub(:get).and_return engine
  end

  describe :handshake do
    let(:message) {{"channel" => "/meta/handshake",
                    "version" => "1.0",
                    "supportedConnectionTypes" => ["long-polling"]
                  }}

    describe "with valid parameters" do
      it "creates a client" do
        engine.should_receive(:create_client)
        server.handshake(message) {}
      end

      it "returns a successful response" do
        engine.stub(:create_client).and_yield "clientid"
        server.handshake(message) do |response|
          response.should == {
            "channel"    => "/meta/handshake",
            "successful" => true,
            "version"    => "1.0",
            "supportedConnectionTypes" => connection_types,
            "clientId"   => "clientid"
          }
        end
      end

      describe "with a message id" do
        before { message["id"] = "foo" }

        it "returns the same id" do
          engine.stub(:create_client).and_yield "clientid"
          server.handshake(message) do |response|
            response.should == {
              "channel"    => "/meta/handshake",
              "successful" => true,
              "version"    => "1.0",
              "supportedConnectionTypes" => connection_types,
              "clientId"   => "clientid",
              "id"         => "foo"
            }
          end
        end
      end
    end

    describe "missing version" do
      before { message.delete "version" }

      it "does not create a client" do
        engine.should_not_receive(:create_client)
        server.handshake(message) {}
      end

      it "returns an unsuccessful response" do
        server.handshake(message) do |response|
          response.should == {
            "channel"    => "/meta/handshake",
            "successful" => false,
            "error"      => "402:version:Missing required parameter",
            "version"    => "1.0",
            "supportedConnectionTypes" => connection_types
          }
        end
      end
    end

    describe "missing supportedConnectionTypes" do
      before { message.delete "supportedConnectionTypes" }

      it "does not create a client" do
        engine.should_not_receive(:create_client)
        server.handshake(message) {}
      end

      it "returns an unsuccessful response" do
        server.handshake(message) do |response|
          response.should == {
            "channel"    => "/meta/handshake",
            "successful" => false,
            "error"      => "402:supportedConnectionTypes:Missing required parameter",
            "version"    => "1.0",
            "supportedConnectionTypes" => connection_types
          }
        end
      end
    end

    describe "with no matching supportedConnectionTypes" do
      before { message["supportedConnectionTypes"] = ["iframe", "flash"] }

      it "does not create a client" do
        engine.should_not_receive(:create_client)
        server.handshake(message) {}
      end

      it "returns an unsuccessful response" do
        server.handshake(message) do |response|
          response.should == {
            "channel"    => "/meta/handshake",
            "successful" => false,
            "error"      => "301:iframe,flash:Connection types not supported",
            "version"    => "1.0",
            "supportedConnectionTypes" => connection_types
          }
        end
      end
    end

    describe "with an error" do
      before { message["error"] = "invalid" }

      it "does not createa a client" do
        engine.should_not_receive(:create_client)
        server.handshake(message) {}
      end

      it "returns an unsuccessful response" do
        server.handshake(message) do |response|
          response.should == {
            "channel"    => "/meta/handshake",
            "successful" => false,
            "error"      => "invalid",
            "version"    => "1.0",
            "supportedConnectionTypes" => connection_types
          }
        end
      end
    end
  end
end
